Objects/longobject.c

```diff
@@ -3051,6 +3094,47 @@ PyLong_AsDouble(PyObject *v)
 
 /* Methods */
 
+void
+_PyLong_ExactDealloc(PyObject *obj)
+{
+    assert(PyLong_CheckExact(obj));
+    PyLongObject *op = (PyLongObject *)obj;
+#if PyLong_MAXFREELIST > 0
+    if (!IS_MEDIUM_VALUE(op)) {
+        PyObject_Del(op);
+        return;
+    }
+    struct _Py_long_state *state = get_long_state();
+#ifdef Py_DEBUG
+    assert(state->numfree != -1);
+#endif
+    if (state->numfree >= PyLong_MAXFREELIST)  {
+        PyObject_Del(op);
+        return;
+    }
+    state->numfree++;
+    Py_SET_TYPE(op, (PyTypeObject *)state->free_list);
+    state->free_list = op;
+#else
+    PyObject_Del(op);
+#endif
+}
+
+static void
+long_dealloc(PyObject *op)
+{
+    assert(PyLong_Check(op));
+#if PyLong_MAXFREELIST > 0
+    if (PyLong_CheckExact(op)) {
+        _PyLong_ExactDealloc(op);
+    }
+    else
+#endif
+    {
+        Py_TYPE(op)->tp_free(op);
```

Comment:
This feels like it should maybe be calling the type's `tp_dealloc`, not `tp_free`? Not going to swear to this, but it feels off to me (then again, calling their `tp_dealloc` could recurse)... Don't take this comment as gospel, I'm probably wrong here.

Reply:
The argument passed in should be long or a subclass of long, continued use of `tp_dealloc` has a high probability of leading to unlimited recursive calls.